[Linky] binding not updating everyday

## Expected Behavior
Linky binding doesn't autoupdate automaticly, I need to deactivate and reactivate things to pull data and update my channel.




## Possible Solution
Auto pull update to have last day consumption updated everyday


## Context

openHAB 3.0.0
Build #2050



## Your Environment
Ubuntu server

I got an exception at 5am:
```
2020-12-13 05:00:33.715 [DEBUG] [.linky.internal.handler.LinkyHandler] - updateData
2020-12-13 05:00:33.719 [DEBUG] [.linky.internal.handler.LinkyHandler] - updateDailyData
2020-12-13 05:00:33.726 [DEBUG] [.linky.internal.api.ExpiringDayCache] - getValue from cache "daily cache" is requiring a fresh value
2020-12-13 05:00:33.730 [DEBUG] [.linky.internal.handler.LinkyHandler] - cachedDailyData refresh
2020-12-13 05:00:33.734 [DEBUG] [.linky.internal.handler.LinkyHandler] - getConsumptionData from 2020-11-28 to 2020-12-13
2020-12-13 05:00:33.738 [DEBUG] [.linky.internal.handler.LinkyHandler] - Calling api.getEnergyData
2020-12-13 05:00:34.190 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 2 column 1 path $
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:224) ~[?:?]
        at com.google.gson.Gson.fromJson(Gson.java:888) ~[?:?]
        at com.google.gson.Gson.fromJson(Gson.java:853) ~[?:?]
        at com.google.gson.Gson.fromJson(Gson.java:802) ~[?:?]
        at com.google.gson.Gson.fromJson(Gson.java:774) ~[?:?]
        at org.openhab.binding.linky.internal.api.EnedisHttpApi.getMeasures(EnedisHttpApi.java:226) ~[?:?]
        at org.openhab.binding.linky.internal.api.EnedisHttpApi.getEnergyData(EnedisHttpApi.java:231) ~[?:?]
        at org.openhab.binding.linky.internal.handler.LinkyHandler.getConsumptionData(LinkyHandler.java:346) ~[?:?]
        at org.openhab.binding.linky.internal.handler.LinkyHandler.lambda$0(LinkyHandler.java:95) ~[?:?]
        at org.openhab.binding.linky.internal.api.ExpiringDayCache.refreshValue(ExpiringDayCache.java:104) ~[?:?]
        at org.openhab.binding.linky.internal.api.ExpiringDayCache.getValue(ExpiringDayCache.java:73) ~[?:?]
        at org.openhab.binding.linky.internal.handler.LinkyHandler.updateDailyData(LinkyHandler.java:204) ~[?:?]
        at org.openhab.binding.linky.internal.handler.LinkyHandler.updateData(LinkyHandler.java:183) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 2 column 1 path $
        at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:385) ~[?:?]
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:213) ~[?:?]
        ... 18 more
```
As I suspected, we have an uncaught exception. I will see if the refresh is run again at 11am or not.

--- a/bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/api/EnedisHttpApi.java
+++ b/bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/api/EnedisHttpApi.java
@@ -45,6 +45,7 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
 
 /**
  * {@link EnedisHttpApi} wraps the Enedis Webservice.
@@ -149,7 +150,7 @@ public class EnedisHttpApi {
                 throw new LinkyException("Connection failed step 5");
             }
             connected = true;
-        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+        } catch (InterruptedException | TimeoutException | ExecutionException | JsonSyntaxException e) {
             throw new LinkyException("Error opening connection with Enedis webservice", e);
         }
     }
@@ -160,6 +161,7 @@ public class EnedisHttpApi {
 
     public void disconnect() throws LinkyException {
         if (connected) {
+            logger.debug("Logout process");
             try { // Three times in a row to get disconnected
                 String location = getLocation(httpClient.GET(URL_APPS_LINCS + "/logout"));
                 location = getLocation(httpClient.GET(location));
@@ -171,6 +173,10 @@ public class EnedisHttpApi {
                 throw new LinkyException("Error while disconnecting from Enedis webservice", e);
             }
         }
+    }
+
+    public boolean isConnected() {
+        return connected;
     }
 
     public void dispose() throws LinkyException {
@@ -204,16 +210,40 @@ public class EnedisHttpApi {
     }
 
     public PrmInfo getPrmInfo() throws LinkyException {
+        if (!connected) {
+            initialize();
+        }
         final String prm_info_url = URL_APPS_LINCS + "/mes-mesures/api/private/v1/personnes/null/prms";
         String data = getData(prm_info_url);
-        PrmInfo[] prms = gson.fromJson(data, PrmInfo[].class);
-        return prms[0];
+        if (data.isEmpty()) {
+            throw new LinkyException(String.format("Requesting '%s' returned an empty response", prm_info_url));
+        }
+        try {
+            PrmInfo[] prms = gson.fromJson(data, PrmInfo[].class);
+            return prms[0];
+        } catch (JsonSyntaxException e) {
+            logger.debug("invalid JSON response not matching PrmInfo[].class: {}", data);
+            throw new LinkyException(String.format("Requesting '%s' returned an invalid JSON response : %s",
+                    prm_info_url, e.getMessage()), e);
+        }
     }
 
     public UserInfo getUserInfo() throws LinkyException {
+        if (!connected) {
+            initialize();
+        }
         final String user_info_url = URL_APPS_LINCS + "/userinfos";
         String data = getData(user_info_url);
-        return Objects.requireNonNull(gson.fromJson(data, UserInfo.class));
+        if (data.isEmpty()) {
+            throw new LinkyException(String.format("Requesting '%s' returned an empty response", user_info_url));
+        }
+        try {
+            return Objects.requireNonNull(gson.fromJson(data, UserInfo.class));
+        } catch (JsonSyntaxException e) {
+            logger.debug("invalid JSON response not matching UserInfo.class: {}", data);
+            throw new LinkyException(String.format("Requesting '%s' returned an invalid JSON response : %s",
+                    user_info_url, e.getMessage()), e);
+        }
     }
 
     private Consumption getMeasures(String userId, String prmId, LocalDate from, LocalDate to, String request)
@@ -223,15 +253,30 @@ public class EnedisHttpApi {
         String url = String.format(measure_url, userId, prmId, request, from.format(API_DATE_FORMAT),
                 to.format(API_DATE_FORMAT));
         String data = getData(url);
-        ConsumptionReport report = gson.fromJson(data, ConsumptionReport.class);
-        return report.firstLevel.consumptions;
+        if (data.isEmpty()) {
+            throw new LinkyException(String.format("Requesting '%s' returned an empty response", url));
+        }
+        try {
+            ConsumptionReport report = gson.fromJson(data, ConsumptionReport.class);
+            return report.firstLevel.consumptions;
+        } catch (JsonSyntaxException e) {
+            logger.debug("invalid JSON response not matching ConsumptionReport.class: {}", data);
+            throw new LinkyException(
+                    String.format("Requesting '%s' returned an invalid JSON response : %s", url, e.getMessage()), e);
+        }
     }
 
     public Consumption getEnergyData(String userId, String prmId, LocalDate from, LocalDate to) throws LinkyException {
+        if (!connected) {
+            initialize();
+        }
         return getMeasures(userId, prmId, from, to, "energie");
     }
 
     public Consumption getPowerData(String userId, String prmId, LocalDate from, LocalDate to) throws LinkyException {
+        if (!connected) {
+            initialize();
+        }
         return getMeasures(userId, prmId, from, to, "pmax");
     }
 }

--- a/bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
+++ b/bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
@@ -356,7 +356,7 @@ public class LinkyHandler extends BaseThingHandler {
             try {
                 return api.getEnergyData(userId, prmId, from, to);
             } catch (LinkyException e) {
-                logger.debug("getConsumptionData failed: {}", e.getMessage(), e);
+                logger.debug("getConsumptionData failed: {}", e.getMessage());
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
             }
         }
@@ -371,7 +371,7 @@ public class LinkyHandler extends BaseThingHandler {
             try {
                 return api.getPowerData(userId, prmId, from, to);
             } catch (LinkyException e) {
-                logger.debug("getPowerData failed: {}", e.getMessage(), e);
+                logger.debug("getPowerData failed: {}", e.getMessage());
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
             }
         }

--- a/bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
+++ b/bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
@@ -356,7 +356,6 @@ public class LinkyHandler extends BaseThingHandler {
             try {
                 return api.getEnergyData(userId, prmId, from, to);
             } catch (LinkyException e) {
-                logger.debug("getConsumptionData failed: {}", e.getMessage());
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
             }
         }
@@ -371,7 +370,6 @@ public class LinkyHandler extends BaseThingHandler {
             try {
                 return api.getPowerData(userId, prmId, from, to);
             } catch (LinkyException e) {
-                logger.debug("getPowerData failed: {}", e.getMessage());
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
             }
         }

--- a/bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
+++ b/bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java
@@ -127,18 +127,25 @@ public class LinkyHandler extends BaseThingHandler {
                     updateStatus(ThingStatus.ONLINE);
 
                     if (thing.getProperties().isEmpty()) {
-                        Map<String, String> properties = discoverAttributes();
+                        Map<String, String> properties = new HashMap<>();
+                        PrmInfo prmInfo = api.getPrmInfo();
+                        UserInfo userInfo = api.getUserInfo();
+                        properties.put(USER_ID, userInfo.userProperties.internId);
+                        properties.put(PUISSANCE, prmInfo.puissanceSouscrite + " kVA");
+                        properties.put(PRM_ID, prmInfo.prmId);
                         updateProperties(properties);
                     }
 
                     prmId = thing.getProperties().get(PRM_ID);
                     userId = thing.getProperties().get(USER_ID);
 
+                    updateData();
+
+                    disconnect();
+
                     final LocalDateTime now = LocalDateTime.now();
                     final LocalDateTime nextDayFirstTimeUpdate = now.plusDays(1).withHour(REFRESH_FIRST_HOUR_OF_DAY)
                             .truncatedTo(ChronoUnit.HOURS);
-
-                    updateData();
 
                     refreshJob = scheduler.scheduleWithFixedDelay(this::updateData,
                             ChronoUnit.MINUTES.between(now, nextDayFirstTimeUpdate) % REFRESH_INTERVAL_IN_MIN + 1,
@@ -152,36 +159,31 @@ public class LinkyHandler extends BaseThingHandler {
         });
     }
 
-    private Map<String, String> discoverAttributes() throws LinkyException {
-        Map<String, String> properties = new HashMap<>();
-        EnedisHttpApi api = this.enedisApi;
-        if (api != null) {
-            PrmInfo prmInfo = api.getPrmInfo();
-            UserInfo userInfo = api.getUserInfo();
-            properties.put(USER_ID, userInfo.userProperties.internId);
-            properties.put(PUISSANCE, prmInfo.puissanceSouscrite + " kVA");
-            properties.put(PRM_ID, prmInfo.prmId);
-        }
-
-        return properties;
-    }
-
     /**
      * Request new data and updates channels
      */
-    private void updateData() {
+    private synchronized void updateData() {
+        boolean connectedBefore = isConnected();
         updatePowerData();
         updateDailyData();
         updateWeeklyData();
         updateMonthlyData();
         updateYearlyData();
+        if (!connectedBefore && isConnected()) {
+            disconnect();
+        }
     }
 
     private synchronized void updatePowerData() {
         if (isLinked(PEAK_POWER) || isLinked(PEAK_TIMESTAMP)) {
             cachedPowerData.getValue().ifPresent(values -> {
-                updateVAChannel(PEAK_POWER, values.aggregats.days.datas.get(0));
-                updateState(PEAK_TIMESTAMP, new DateTimeType(values.aggregats.days.periodes.get(0).dateDebut));
+                Aggregate days = values.aggregats.days;
+                if (days.datas.size() == 0 || days.periodes.size() == 0) {
+                    logger.debug("Daily power data are without any period/data");
+                } else {
+                    updateVAChannel(PEAK_POWER, days.datas.get(0));
+                    updateState(PEAK_TIMESTAMP, new DateTimeType(days.periodes.get(0).dateDebut));
+                }
             });
         }
     }
@@ -193,6 +195,10 @@ public class LinkyHandler extends BaseThingHandler {
         if (isLinked(YESTERDAY) || isLinked(THIS_WEEK)) {
             cachedDailyData.getValue().ifPresent(values -> {
                 Aggregate days = values.aggregats.days;
+                if (days.periodes.size() > days.datas.size()) {
+                    logger.debug("Daily data are invalid: not a data for each period");
+                    return;
+                }
                 int maxValue = days.periodes.size() - 1;
                 int thisWeekNumber = days.periodes.get(maxValue).dateDebut.get(weekFields.weekOfWeekBasedYear());
                 double yesterday = days.datas.get(maxValue);
@@ -223,6 +229,9 @@ public class LinkyHandler extends BaseThingHandler {
                 Aggregate weeks = values.aggregats.weeks;
                 if (weeks.datas.size() > 1) {
                     updateKwhChannel(LAST_WEEK, weeks.datas.get(1));
+                } else {
+                    logger.debug("Weekly data are without last week data");
+                    updateKwhChannel(LAST_WEEK, Double.NaN);
                 }
             });
         }
@@ -235,11 +244,18 @@ public class LinkyHandler extends BaseThingHandler {
         if (isLinked(LAST_MONTH) || isLinked(THIS_MONTH)) {
             cachedMonthlyData.getValue().ifPresent(values -> {
                 Aggregate months = values.aggregats.months;
-                updateKwhChannel(LAST_MONTH, months.datas.get(0));
-                if (months.datas.size() > 1) {
-                    updateKwhChannel(THIS_MONTH, months.datas.get(1));
-                } else {
+                if (months.datas.size() == 0) {
+                    logger.debug("Monthly data are without any data");
+                    updateKwhChannel(LAST_MONTH, Double.NaN);
                     updateKwhChannel(THIS_MONTH, Double.NaN);
+                } else {
+                    updateKwhChannel(LAST_MONTH, months.datas.get(0));
+                    if (months.datas.size() > 1) {
+                        updateKwhChannel(THIS_MONTH, months.datas.get(1));
+                    } else {
+                        logger.debug("Monthly data are without current month data");
+                        updateKwhChannel(THIS_MONTH, Double.NaN);
+                    }
                 }
             });
         }
@@ -252,11 +268,18 @@ public class LinkyHandler extends BaseThingHandler {
         if (isLinked(LAST_YEAR) || isLinked(THIS_YEAR)) {
             cachedYearlyData.getValue().ifPresent(values -> {
                 Aggregate years = values.aggregats.years;
-                updateKwhChannel(LAST_YEAR, years.datas.get(0));
-                if (years.datas.size() > 1) {
-                    updateKwhChannel(THIS_YEAR, years.datas.get(1));
-                } else {
+                if (years.datas.size() == 0) {
+                    logger.debug("Yearly data are without any data");
+                    updateKwhChannel(LAST_YEAR, Double.NaN);
                     updateKwhChannel(THIS_YEAR, Double.NaN);
+                } else {
+                    updateKwhChannel(LAST_YEAR, years.datas.get(0));
+                    if (years.datas.size() > 1) {
+                        updateKwhChannel(THIS_YEAR, years.datas.get(1));
+                    } else {
+                        logger.debug("Yearly data are without current year data");
+                        updateKwhChannel(THIS_YEAR, Double.NaN);
+                    }
                 }
             });
         }
@@ -280,9 +303,15 @@ public class LinkyHandler extends BaseThingHandler {
      * @param endDay the end day of the report
      * @param separator the separator to be used betwwen the date and the value
      *
-     * @return the report as a string
+     * @return the report as a list of string
      */
-    public List<String> reportValues(LocalDate startDay, LocalDate endDay, @Nullable String separator) {
+    public synchronized List<String> reportValues(LocalDate startDay, LocalDate endDay, @Nullable String separator) {
+        List<String> report = buildReport(startDay, endDay, separator);
+        disconnect();
+        return report;
+    }
+
+    private List<String> buildReport(LocalDate startDay, LocalDate endDay, @Nullable String separator) {
         List<String> report = new ArrayList<>();
         if (startDay.getYear() == endDay.getYear() && startDay.getMonthValue() == endDay.getMonthValue()) {
             // All values in the same month
@@ -312,7 +341,7 @@ public class LinkyHandler extends BaseThingHandler {
                 if (last.isAfter(endDay)) {
                     last = endDay;
                 }
-                report.addAll(reportValues(first, last, separator));
+                report.addAll(buildReport(first, last, separator));
                 first = last.plusDays(1);
             } while (!first.isAfter(endDay));
         }
@@ -320,11 +349,14 @@ public class LinkyHandler extends BaseThingHandler {
     }
 
     private @Nullable Consumption getConsumptionData(LocalDate from, LocalDate to) {
+        logger.debug("getConsumptionData from {} to {}", from.format(DateTimeFormatter.ISO_LOCAL_DATE),
+                to.format(DateTimeFormatter.ISO_LOCAL_DATE));
         EnedisHttpApi api = this.enedisApi;
         if (api != null) {
             try {
                 return api.getEnergyData(userId, prmId, from, to);
             } catch (LinkyException e) {
+                logger.debug("getConsumptionData failed: {}", e.getMessage(), e);
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
             }
         }
@@ -332,15 +364,33 @@ public class LinkyHandler extends BaseThingHandler {
     }
 
     private @Nullable Consumption getPowerData(LocalDate from, LocalDate to) {
+        logger.debug("getPowerData from {} to {}", from.format(DateTimeFormatter.ISO_LOCAL_DATE),
+                to.format(DateTimeFormatter.ISO_LOCAL_DATE));
         EnedisHttpApi api = this.enedisApi;
         if (api != null) {
             try {
                 return api.getPowerData(userId, prmId, from, to);
             } catch (LinkyException e) {
+                logger.debug("getPowerData failed: {}", e.getMessage(), e);
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
             }
         }
         return null;
+    }
+
+    private boolean isConnected() {
+        EnedisHttpApi api = this.enedisApi;
+        return api == null ? false : api.isConnected();
+    }
+
+    private void disconnect() {
+        EnedisHttpApi api = this.enedisApi;
+        if (api != null) {
+            try {
+                api.dispose();
+            } catch (LinkyException ignore) {
+            }
+        }
     }
 
     @Override
@@ -351,25 +401,22 @@ public class LinkyHandler extends BaseThingHandler {
             job.cancel(true);
             refreshJob = null;
         }
-        EnedisHttpApi api = this.enedisApi;
-        if (api != null) {
-            try {
-                api.dispose();
-                enedisApi = null;
-            } catch (LinkyException ignore) {
-            }
-        }
+        disconnect();
+        enedisApi = null;
     }
 
     @Override
-    public void handleCommand(ChannelUID channelUID, Command command) {
+    public synchronized void handleCommand(ChannelUID channelUID, Command command) {
         if (command instanceof RefreshType) {
             logger.debug("Refreshing channel {}", channelUID.getId());
+            boolean connectedBefore = isConnected();
             switch (channelUID.getId()) {
                 case YESTERDAY:
-                case LAST_WEEK:
                 case THIS_WEEK:
                     updateDailyData();
+                    break;
+                case LAST_WEEK:
+                    updateWeeklyData();
                     break;
                 case LAST_MONTH:
                 case THIS_MONTH:
@@ -379,8 +426,15 @@ public class LinkyHandler extends BaseThingHandler {
                 case THIS_YEAR:
                     updateYearlyData();
                     break;
+                case PEAK_POWER:
+                case PEAK_TIMESTAMP:
+                    updatePowerData();
+                    break;
                 default:
                     break;
+            }
+            if (!connectedBefore && isConnected()) {
+                disconnect();
             }
         } else {
             logger.debug("The Linky binding is read-only and can not handle command {}", command);

